packages/snowpack/src/logger.ts

```diff
@@ -1,35 +1,88 @@
+const levels: Record<LoggerLevel, number> = {
+  debug: 20,
+  info: 30,
+  warn: 40,
+  error: 50,
+  silent: 90,
+};
 
-    if (inputData.level === LEVEL.ERROR || inputData.level === LEVEL.FATAL) msg = colors.red(msg);
-    if (inputData.level === LEVEL.WARN) msg = colors.yellow(msg);
+/** Custom logger heavily-inspired by https://github.com/pinojs/pino with extra features like log retentian */
+class SnowpackLogger {
+  /** set the log level (can be changed after init) */
+  public level: LoggerLevel = 'info';
 
-    return msg;
+  private history: string[] = []; // this is immutable; must be accessed with Logger.getHistory()
+  private callbacks: Record<LoggerEvent, (message: string) => void> = {
+    debug: () => {},
+    info: () => {},
+    warn: () => {},
+    error: () => {},
   };
-}
 
-/** use Pino, a tiny logging library */
-export default function createLogger(opts: pino.LoggerOptions = {}) {
-  return pino({
-    ...opts,
-    name: opts.name || 'snowpack',
-    prettyPrint: {suppressFlushSyncWarning: true},
-    prettifier,
-  });
+  private log({level, name, message}: {level: LoggerEvent; name: string; message: string}) {
+    // test if this level is enabled or not
+    if (levels[this.level] > levels[level]) {
+      return; // do nothing
+    }
+
+    // format
+    let text = message;
+    if (level === 'warn') text = colors.yellow(text);
+    if (level === 'error') text = colors.red(text);
+    const log = `${colors.dim(`[${name}]`)} ${text}`;
+
+    // add to log history
+    this.history = this.history.concat(log);
+
+    // log
+    let logFn = console.log;
+    if (level === 'warn') logFn = console.warn;
+    if (level === 'error') logFn = console.error;
+    logFn(log);
+
+    // fire callback, if any
+    if (typeof this.callbacks[level] === 'function') {
+      this.callbacks[level](log);
+    }
+  }
+
+  /** emit messages only visible when --debug is passed */
+  public debug(message: string, options?: LoggerOptions): void {
+    const name = (options && options.name) || 'snowpack';
+    this.log({level: 'debug', name, message});
+  }
+
+  /** emit general info */
+  public info(message: string, options?: LoggerOptions): void {
+    const name = (options && options.name) || 'snowpack';
+    this.log({level: 'info', name, message});
+  }
+
+  /** emit non-fatal warnings */
+  public warn(message: string, options?: LoggerOptions): void {
+    const name = (options && options.name) || 'snowpack';
+    this.log({level: 'warn', name, message});
+  }
+
+  /** emit critical error messages */
+  public error(message: string, options?: LoggerOptions): void {
+    const name = (options && options.name) || 'snowpack';
+    this.log({level: 'error', name, message});
+  }
+
+  /** get full logging history */
+  public getHistory() {
+    return [...this.history];
+  }
+
+  /** listen for events */
+  public on(event: LoggerEvent, callback: (message: string) => void) {
+    this.callbacks[event] = callback;
```

Comment:
So I’ll admit: I had a bit of a brainfart when it came time to implement this. The API I was shooting for was:

```js
logger.on('warn', (message) => {
  // do something with message
});
```

I admittedly may have implemented this a dumb way; open to feedback if so.

Reply:
Yea, based on the usage below it looks like we don't need multiple handlers based on log level, and just need `logger.on((message) => { ...`

Reply:
I actually like the API; I more-or-less meant the class implementation. I feel like if we did `logger.on(message => {` we’d need a 2nd param for what type of log it is, e.g. `info` vs `debug` vs `error`. There’s extra context needed to handle different levels differently, and that’s lost with a generic callback.

Also, see the update pushed, where now calling `on('error') => ` actually overwrites the default behavior, so we’d need that log level context too.